### ISSUE TITLE
Audible Notification of Spelling/Grammer Errors: Allow user to Toggle Setting

### ISSUE THREAD

**Issue by elliott94** [nvaccess/nvda#5872]
Whilst the new ability for NVDA to detect potential spelling errors on-the-fly is a nice feature, it'd be beneficial to allow this setting to be toggled if the user does not wish for the information to be conveyed whilst typing. In addition to a sound-only approach, as is currently possible for capital letters presence of these errors could also be indicated using speech - by announcing errors in a lower pitch, for example.

**Comment by jcsteh** [comment]
You can already disable this by turning off reporting of spelling 
errors. If you have reporting of spelling errors enabled, that suggests 
you want to know about them; I can't think of any use case where you 
wouldn't want to know about them sooner.

There are two problems with speaking this instead of using a sound:
1. Your typing will probably interrupt the speech about the error, so 
you may not actually hear it.
2. This is meant to be a warning you can very quickly notice without 
necessarily interrupting what you're doing. Speech, on the other hand, 
requires more concentration because you have to pay attention to what is 
being spoken, distracting you from the task on which you are primarily 
focused (typing text).

**Comment by elliott94** [comment]
Ah, thanks for the tip - I should have tried disabling this beforehand.

Still, though, from a personal point-of-view I much prefer only being notified about spelling errors as I manually review a document - for example, I might be using lots of technical words that Office may not necesarily have dictionary definitions for. Would you perhaps just consider allowing the user to change whether a sound is played, essentially reverting back to the old behaviour?

**Comment by jcsteh** [comment]
The thing is that if you have this enabled, you're still going to be 
"distracted" by spelling errors you don't care about when, for example, 
reading the current line or word. If you really don't want to know about 
them until later, why not just disable spelling errors until you're 
ready for review? You've clearly already separated your workflow into 
two parts: entry and review.

**Comment by elliott94** [comment]
I could do this - however, within Word there are several phrases that when typed generate errors before they're corrected. For example, when typing the word "they're", after entering the apostrophy but before adding the "re", the error sound is played which can get somewhat annoying.

That said, however, I understand why the announcement of errors whilst typing wouldn't be a good thing - I just think that it'd be nice to have the option of either announcing spelling errors whilst reviewing the document in conjunction with a sound, or either or neither of those.

**Comment by jcsteh** [comment]
In 414fdc5d, errors are no longer indicated when apostrophe is pressed. I do agree this was annoying and counter-productive.

Aside from this, though, I'm not seeing a good use case for separating these options.

**Comment by derekriemer** [comment]
Did you add apostrophe to this as well?

Sent from my heavily encrypted iPhone.
Please disregard errors as this is a smaller device.

**Comment by derekriemer** [comment]
Never mind my post here, my stupid phone read that wrong (or my brain 
heard it wrong).

On 4/11/2016 6:39 PM, James Teh wrote:
>
> In 414fdc5 
> <https://github.com/nvaccess/nvda/commit/414fdc5d23f96413fd0b4ff81c3fa98eef9140eb>, 
> errors are no longer indicated when apostrophe is pressed. I do agree 
> this was annoying and counter-productive.
>
> Aside from this, though, I'm not seeing a good use case for separating 
> these options.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5872#issuecomment-208643064>
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by elliott94** [comment]
Thank you for making the change. :)

For me, it's the idea of flexibility - in my view, there are several ways of checking input before it's processed:

1. On-The-Fly - With this new feature, I can understand how the process of checking documents can now in theory be completed much quicker than was previously possible - if a user is typing and they hear the notification, if desired they can simply edit the text and continue.

2. Complete Review - After a user has finished typing information, they can then either use the built-in spelling tools of the software, or manually go through the document and get notified by NVDA if any errors are present within the text and edit/ignore as necessary.

For me, the second approach is usually what I go for - having the ability to instantly receive feedback on potential errors whilst reviewing a document (rather than when I'm typing) is paramount for me. Strangely, however, after several days of real-world usage I've actually found the sound somewhat annoying. :) Whilst it's true that I've been able to spot errors much quicker, rather than being constantly notified that I may have accidentally mistyped a word, I personally prefer to simply read a completed document when finished, and make any adjustments that I feel should be made. I simply can't justify the time taken to disable the announcements of spelling errors at the start of a potentially long typing session, only to re-enable them afterwards rather than simply performing an entire spelling check on the document.

However, getting back to my original point... I've just taken a look at the contents of the waves directory, and with exception of the error sound (which justifiably can't be disabled), I'm able to toggle all other sound events. Does it not make sense to allow this for spelling errors also?

**Comment by jcsteh** [comment]
I don't quite follow your logic. You say "I simply can't justify the 
time taken to disable the announcements of spelling errors at the start 
of a potentially long typing session", but if it's a long session, the 
time taken to change the setting is negligible in comparison to the time 
spent typing. Also, are you suggesting you *never* naviate through text 
at all as you are typing? If you do navigate at all, you will hear about 
spelling errors, which means you're hearing about them before you do a 
full review anyway, thus breaking your requirement that you not learn 
about errors until your later review.

It's also worth noting that you can assign a single key to toggle 
reporting of spelling errors.

**Comment by elliott94** [comment]
@jcsteh said:

> I don't quite follow your logic. You say "I simply can't justify the
time taken to disable the announcements of spelling errors at the start
of a potentially long typing session", but if it's a long session, the
time taken to change the setting is negligible in comparison to the time
spent typing.

In hindsight, I realise that I could have justified myself better. I realise that I could assign a custom jesture to this, but the point that I was trying to make was that whilst it's entirely possible to disable this, with the current implementation it's not possible to have it both ways. If a user who is currently used to the old behaviour of verbal announcement of errors upgrades to the new version, the new method of audible notifications whilst typing has the potential to get somewhat disconcerting if only speech-based notification is desired.

@jcsteh said:

> Also, are you suggesting you *never* naviate through text
at all as you are typing? If you do navigate at all, you will hear about
spelling errors, which means you're hearing about them before you do a
full review anyway, thus breaking your requirement that you not learn
about errors until your later review.

If I'm working on an email, for example, if I know for a fact that I've made a mistake whilst typing then I'll immediately go back and make the required change - ultimately though, it's all about what I'm typing at the time and the length of the document.

In summary, the point I'm trying to make is this. NVDA has always been about flexibility - if a user wishes to know whether the text they're focused on is capitalised, they have the option of a beep, a verbal pitch change, or both - the same can be said for progress bars. Whilst this issue may seem trivial, I just think that having the option available to users gives them the choice on how they receive that information rather than only having one behaviour available.

**Comment by jcsteh** [comment]
Flexibility is a good thing, but too many options leads to confusion and 
an overwhelming interface. There should only be an option for something 
if there's actually a solid use case for it. Use cases come down to intent.

The thing I'm not following here is that if you're intending to know 
about spelling errors, it makes sense that you'd want to... know about 
spelling errors. The option isn't specific to review. You're saying you 
just want the ability to choose, but I'm not following the 
*disadvantage* of knowing earlier. We don't provide the choice between 
speech and audio in this case because speech wouldn't work for this 
indication.

**Comment by elliott94** [comment]
I fully agree with your observation that speech wouldn't work in this use case - I realise that it's something that I speculated in my initial issue text, but you're right above when you say that it just wouldn't be feasible to add considering that any warnings would most likely be missed anyway.

For me, it's all about how the information is sent by NVDA. I like being able to quickly browse through text and instantly get notification that a certain block may be incorrect - however, I'm just not liking the audible notification whilst typing - I want to focus on each word that I type, rather than the distraction of a new sound - if I know for a fact that I've mistyped a word which will become apparent as soon as I hit my space bar, I just don't find the new method useful. That's not to say, however, that I would want to disable the functionality altogether just to re-enable it after I've finished typing - I realise that I'm mixing and matching review styles here! :)

**Comment by PratikP1** [comment]
As I’m reviewing text, it bothers me more when a screen reader constantly tells me that some  block of text is incorrect, grammar, spelling, etc. This is especially true when I don’t trust Microsoft’s judgment on what constitutes bad grammar. I explicitly turn off grammar checking in spellcheck. It appears, however, that NVDA does not have a way to honor this setting. When I’m writing, I want to focus on content more than the technical editing that comes after the writing is done. Spelling notifications are fine as I’m writing. Those are easy to fix or ignore. Grammatical mistakes are not. Spending time fixing mistakes takes me out of the writing flow. I would rather not have those interruptions every time I’m reviewing text for content.

**Comment by jcsteh** [comment]
I don't quite follow. You can easily disable reporting of spelling errors, etc. by unchecking, for example, Report spelling errors in NVDA's Document Formatting settings. As you say you don't care about these errors, disabling this is what you want.

**Comment by PratikP1** [comment]
Yes. you’re absolutely right. It also switches off the spelling error tone. Also switches off the grammar announcement which is far more disruptive. Unless there is a technical reason for not doing so, these two should be separated as settings. There should be one for grammar and another one for spelling.

**Comment by jcsteh** [comment]
Separating grammar is a fair argument and I think we should do this. However, that's a different argument to separating spelling errors while typing.

**Comment by PratikP1** [comment]
If grammar and spelling are separated, then the problem of spelling detection while typing becomes a user choice. The user can enable or disable the spelling detection setting. I have no problem with that.

**Comment by jcsteh** [comment]
@MichaelDCurran, @Qchristensen, I'd love your thoughts on the below:

1. Should we have a separate option for indication of spelling errors as you type (which is done with a sound)? Currently, it's controlled by the "Report spelling errors" option. There's plenty of discussion in earlier comments, but briefly, I argue that if your intent is to know about spelling errors, hearing the indication as you type fits that intent. @elliott94 argues that he only wants to know about them when reviewing later, but in that case, the intent is different, so one should just disable it while typing and enable it when reviewing.
2. If we *do* have a separate option, where should it go? I think it makes more sense in Keyboard Settings than Document Formatting, but it's tricky. If we did put it in Document Formatting, it'd be somewhat different to all of the other options, since this one is specific to typing. On the flip side, it might confuse people if it goes in Keyboard Settings.
3. Again, if we *do* have a separate option, should it only be valid when reporting of spelling errors (the existing option) is enabled? I believe so, since hearing an alert for errors and then not hearing about them when looking back to find the error is counter-intuitive at best. However, if the option goes in Keyboard Settings, this gets a bit ugly.

**Comment by derekriemer** [comment]
@jcsteh @MichaelDCurran @Qchristinsen How would a combo box for spelling 
errors that can report the following work?
Report spelling errors
* Off
*While reading
* While reading and typing.

The disadvantage of this is that existing configs will need migrated 
(right now it's a boolean value. We'd need an int or choice in the config.
The advantage would be putting this in one control which makes sense to me.


On 4/15/2016 6:06 AM, James Teh wrote:
>
> @MichaelDCurran <https://github.com/MichaelDCurran>, @Qchristensen 
> <https://github.com/Qchristensen>, I'd love your thoughts on the below:
>
>  1. Should we have a separate option for indication of spelling errors
>     as you type (which is done with a sound)? Currently, it's
>     controlled by the "Report spelling errors" option. There's plenty
>     of discussion in earlier comments, but briefly, I argue that if
>     your intent is to know about spelling errors, hearing the
>     indication as you type fits that intent. @elliott94
>     <https://github.com/elliott94> argues that he only wants to know
>     about them when reviewing later, but in that case, the intent is
>     different, so one should just disable it while typing and enable
>     it when reviewing.
>  2. If we /do/ have a separate option, where should it go? I think it
>     makes more sense in Keyboard Settings than Document Formatting,
>     but it's tricky. If we did put it in Document Formatting, it'd be
>     somewhat different to all of the other options, since this one is
>     specific to typing. On the flip side, it might confuse people if
>     it goes in Keyboard Settings.
>  3. Again, if we /do/ have a separate option, should it only be valid
>     when reporting of spelling errors (the existing option) is
>     enabled? I believe so, since hearing an alert for errors and then
>     not hearing about them when looking back to find the error is
>     counter-intuitive at best. However, if the option goes in Keyboard
>     Settings, this gets a bit ugly.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub 
> <https://github.com/nvaccess/nvda/issues/5872#issuecomment-210439217>
>

-- 
------------------------------------------------------------------------


    Derek Riemer

  * Department of computer science, third year undergraduate student.
  * Proud user of the NVDA screen reader.
  * Open source enthusiast.
  * Member of Bridge Cu
  * Avid skiier.

Websites:
Honors portfolio <http://derekriemer.com>
Awesome little hand built weather app! 
<http://django.derekriemer.com/weather/>

email me at derek.riemer@colorado.edu <mailto:derek.riemer@colorado.edu>
Phone: (303) 906-2194

**Comment by elliott94** [comment]
Thank you for adding this option. :) As I mentioned previously, this is one of the main things that I love about NVDA - even though you (@jcsteh) didn't agree with a separate option and its use, you were willing to openly discuss and add it as a feature.
